pkg/deps/resolve.go

```diff
@@ -0,0 +1,167 @@
+	c := r.componentLookup[ref]
+	if c == nil {
+		return nil
+	}
+	return c.DeepCopy()
+}
+
+// HasComponent indicates whether a component is known by the resolver.
+func (r *Resolver) HasComponent(ref bundle.ComponentReference) bool {
+	return r.componentLookup[ref] != nil
+}
+
+// Add adds components to the resolver, returning a new resolver or
+// an error if it fails. If one of the new components provided has the same
+// name/version, it overwrites the existing component map.
+func (r *Resolver) Add(newComps []*bundle.Component) (*Resolver, error) {
+	m := make(map[bundle.ComponentReference]*bundle.Component)
+	for k, c := range r.componentLookup {
+		m[k] = c
+	}
+	for _, c := range newComps {
+		m[c.ComponentReference()] = c
+	}
+	var cs []*bundle.Component
+	for _, c := range m {
+		cs = append(cs, c)
+	}
+	return NewResolver(cs)
+}
+
+// ComponentVersions gets the versions for a particular component
+func (r *Resolver) ComponentVersions(comp string) ([]bundle.ComponentReference, error) {
+	cv, ok := r.componentVersions[comp]
+	if !ok {
+		return nil, fmt.Errorf("unknown component %q", comp)
+	}
+	var versions []bundle.ComponentReference
+	for _, ver := range cv.versions {
+		versions = append(versions, bundle.ComponentReference{
+			ComponentName: cv.componentName,
+			Version:       ver.version.String(),
+		})
+	}
+	return versions, nil
+}
+
+// ResolveOptions are options for resolving dependencies
+type ResolveOptions struct {
+	// Match are component annotations that must to be present. This is useful
+	// for matching positive characteristics of a component (example: this
+	// component passed qualification). Only one of the list of values need
+	// be present.
+	//
+	// In other words, this is a logical AND operation of the passed in Annotation and
+	// the component Annotation.
+	Match map[string][]string
+
+	// MatchIfPresent are component annotations that are analyzed if they are
+	// present on the component. This is useful for synchronizing values between
+	// the universe of components and the caller.
+	MatchIfPresent map[string][]string
+
+	// Exclude are component annotations that must not be present.  This is
+	// useful for matching positive characteristics of a component (this
+	// component passed qualification). If any of the the list of values match,
+	// the component is excluded.
+	//
+	// In other words, this is a logical NAND operation of the passed in
+	// Annotation and the component Annotation.
+	Exclude map[string][]string
+}
+
+// ResolveLatest resolves dependencies for several components, returning
+// the components references that correspond to that version selection.
+//
+// These components must exist within the Resolver's set of components. If a
+// version is not specified for a component reference, the latest version is
+// used.
+func (r *Resolver) ResolveLatest(refs []bundle.ComponentReference, opts *ResolveOptions) ([]bundle.ComponentReference, error) {
```

Comment:
'Latest' is a bit of a misnomer, as some versions may be downgraded from latest.